crates/wasmparser/src/collections/index_set.rs

```diff
@@ -0,0 +1,241 @@
+//! Type definitions for an ordered set.
+
+use core::{borrow::Borrow, hash::Hash, iter::FusedIterator, ops::Index};
+
+#[cfg(not(feature = "no-hash-maps"))]
+mod detail {
+    use crate::collections::hash;
+
+    pub type IndexSetImpl<T> = indexmap::IndexSet<T, hash::RandomState>;
+    pub type IterImpl<'a, T> = indexmap::set::Iter<'a, T>;
+    pub type IntoIterImpl<T> = indexmap::set::IntoIter<T>;
+}
+
+#[cfg(feature = "no-hash-maps")]
+mod detail;
```

Comment:
Would it make sense to build this module on `crate::collections::IndexMap<T, ()>` instead of having this `detail` split as well?

Reply:
I will try it out and do so if it is simple.

Reply:
I tried very hard to implement this, however, especially the `IndexSet::replace` method is very hard to properly implement in terms of `IndexMap<T, ()>` since it requires to change the underlying keys which is not possible directly with `IndexMap`. For this the original `IndexSet` data structure uses the low-level underlying `RawMap` which we cannot simply do. Also tests indicate that even the current `replace` API is bugged since for similar reasons it clones the key instead of actually replacing it. I thought it was enough but tests mandate a change. I am not yet sure how to fix this with the current architecture tbh.

Reply:
If `replace` is used rarely then it might be an option to implement that as a simple remove-then-insert perhaps? That should at least semantically be correct and it could always be sped up in the future if needed